Add support for referencing forge's resources and specifying existing mods to data generators

pupnewfster opened the pull request:
This PR adds support for declaring that a mod's resources should be added to the existing file helper's resource check b adding an `--existing-mod` argument that takes a modid. This PR also auto-adds forge as an existing mod so that modders can reference forge resources from their data generators without having running into errors, similar to how vanilla resources can be referenced.

pupnewfster committed:
Add --existing-mod as an argument for data generators that takes a modid and auto add forge so that mods can reference forge resources

cpw commented:
Does this affect forge's datagen? We use this to generate forge's resources too..

pupnewfster commented:
It seems to mostly work.

The way I tested it was by adding the line I am changing in the data generator test and validating it crashed. Then I added forge as existing by specifying it with the new `--existing-mod` argument, and it properly worked. Next I removed the manual specification of forge existing and uncommented to make forge be automatically added (similar to vanilla being added), which still worked. The reason I believe it works for forge's own data generators still is that I also at the end tested by running forge's data generator and it ran fine, didn't end up changing any of the files or crash.

I also just tested by removing some of forge's generated resources to see if it properly is unable to find them on the initial run or if it thinks they exist for one run before it then deletes them and it seems to find them. Do you think a reasonable solution to this would be to make it so that forge only gets automatically added if it has not been specified as a `--mod`?

pupnewfster committed:
Only automatically add forge to existing resources if forge is not one of the mods data is currently being generated for

TheCurle commented on patches/minecraft/net/minecraft/data/Main.java.patch:
```diff
@@ -24,9 +25,10 @@
 +         java.util.Set<String> mods = new java.util.HashSet<>(optionset.valuesOf(mod));
 +         boolean isFlat = mods.isEmpty() || optionset.has(flat);
-+         net.minecraftforge.fml.DatagenModLoader.begin(mods, path, inputs, existingPacks, flag2, flag1, flag3, flag4, flag5, isFlat);
++         net.minecraftforge.fml.DatagenModLoader.begin(mods, path, inputs, existingPacks, existingMods, flag2, flag1, flag3, flag4, flag5, isFlat);
```
Why is this in the middle of the argument list?

TheCurle commented on src/main/java/net/minecraftforge/fml/DatagenModLoader.java:
```diff
@@ -39,14 +40,19 @@ public static boolean isRunningDataGen() {
+        begin(mods, path, inputs, existingPacks, Collections.emptySet(), serverGenerators, clientGenerators, devToolGenerators, reportsGenerator, structureValidator, flat);
+    }
+
+    public static void begin(final Set<String> mods, final Path path, final Collection<Path> inputs, Collection<Path> existingPacks, Set<String> existingMods, final boolean serverGenerators, final boolean clientGenerators, final boolean devToolGenerators, final boolean reportsGenerator, final boolean structureValidator, final boolean flat) {
```
Middle of the argument list..
This is a Forge class yes, but why?

pupnewfster replied:
Because argument wise it makes more sense going next to existing packs than at the end after a ton of random boolean values, and once we hit 1.17 and the deprecated old one can be removed without worrying that someone is calling it for some reason from somewhere else, it really won't matter as it won't be the "middle" it will just be the "normal" place for it.

TheCurle reviewed (commented):
I'm going to test this a little. 
Looks okay, apart from some weirdness with the API changes.

pupnewfster committed:
Cleanup
